File: TwoPointer/1424.diagonal-traverse-ii.go
package leetcode

import "slices"

// @leet start
func findDiagonalOrder(nums [][]int) []int {
	m := map[int][]int{}
	var n int
	for i, row := range nums {
		for j := 0; j < len(row); j++ {
			m[i+j] = append(m[i+j], row[j])
			n = max(n, i+j)
		}
	}
	res := make([]int, 0, n)
	for i := range n + 1 {
		cur := m[i]
		slices.Reverse(cur)
		res = append(res, cur...)
	}
	return res
}

// @leet end

File: TwoPointer/1424.diagonal-traverse-ii.py
from collections import defaultdict
from typing import List


# @leet start
class Solution:
    def findDiagonalOrder(self, nums: List[List[int]]) -> List[int]:
        res = []
        n = 0
        diags = defaultdict(list)
        for i in range(len(nums)):
            row = nums[i]
            for j in range(len(row)):
                diags[i + j].append(row[j])
                n = max(n, i + j)

        for i in range(n + 1):
            row = diags[i]
            row.reverse()
            res.extend(row)
        return res


# @leet end

